### PR TITLE
added interactive snow shader

### PR DESCRIPTION
**Pull request by QbieShay**
**Please check if the PR fulfills these requirements:**

- [ ] The commit message follows our guidelines.
- For bug fixes and features:
    - [ ] You tested the changes.
    - [ ] You updated the docs or changelog.

**What kind of change does this PR introduce?**
An interactive snow shader.
The splatmap is painted dynamically in a separate Viewport, and it tracks the movement of an object in the scene.
This is used in the demo to blend between snow and dirt, but can be used to blend any type of textures.

**Does this PR introduce a breaking change?**
No, it adds a new demo.

## New feature or change ##

**What is the current behavior?** 
No interactive snow demo

**What is the new behavior?**
New interactive snow demo!

**Other information**
image

### CODE REVIEW

File: godot/Shaders/interactive_snow.shader [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+uniform sampler2D blend_texture;
+uniform float snow_height = 0.3;
+uniform float snow_sharpness: hint_range(0.3, 20.0, 0.1) = 3.0;
+uniform vec2 textures_tiling = vec2(4.0);
+
+varying vec2 world_uv;
+varying float height;
+
+void vertex(){
+
```
Comment (aaronfranke) [inline godot/Shaders/interactive_snow.shader:22]:
```suggestion
```

File: godot/Shaders/interactive_snow.shader [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	// The coordinates need to be divided by the in-world size of the texture 
+	world_uv /= mask_texture_world_size;
+	// world coordinates can be negative, so we need to offset them by 0.5 to bring them in the (0,1) range
+	world_uv += vec2(0.5);
+	// The next part is about calculating the normal of the face.
+	// Because normal, tangent and binormal are related to the variation of the y along
+	// an axis, we need to first establish how small of an increment we want to consider.
+	// Here we take the size of the world texture, and divide it by the texture size.
+
+	vec2 texture_increment = vec2(mask_texture_world_size)/ vec2(textureSize(mask_texture, 0));
```
Comment (aaronfranke) [inline godot/Shaders/interactive_snow.shader:35]:
```suggestion
	vec2 texture_increment = vec2(mask_texture_world_size) / vec2(textureSize(mask_texture, 0));
```

File: godot/Shaders/interactive_snow.shader [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		mask * noise_interp * 2.0,
+		1.0 - 2.0 * (1.0 - mask) * (1.0 - noise_interp),
+		step(0.5, mask)
+	);
+	snow_interp = 1.0 - pow(1.0 - snow_interp, snow_sharpness);
+	snow_interp = clamp(snow_interp, 0.0, 1.0);
+	ALBEDO = mix(dirt_color.rgb, snow_color.rgb, snow_interp);
+	NORMALMAP = mix(dirt_normal, snow_normal, snow_interp);
+	ROUGHNESS = mix(dirt_roughness, snow_rougness, snow_interp);
+}
```
Comment (aaronfranke) [inline godot/Shaders/interactive_snow.shader:72]: Missing newline at EOF.

File: godot/Demos/InteractiveSnow/InteractiveSnowball.gd [hunk 4]
```diff
@@ -0,0 +1,16 @@
+extends MeshInstance
+
+export var move_speed = 5.0
+export var rotate_speed = 180.0
+export var trail_sprite_path: NodePath = ".."
+export var interactive_snow_plane_path: NodePath = "../InteractiveSnowPlane/Plane"
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowball.gd:6]: `@` indicates a NodePath literal.

```suggestion
export var trail_sprite_path: NodePath = @".."
export var interactive_snow_plane_path: NodePath = @"../InteractiveSnowPlane/Plane"
```

File: godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	var line = get_node("Line2D")
+	line.add_point( (player_pos/ interactive_snow_size) * size)
+
+func _process(delta):
+	var player = get_node(interactive_actor_path)
+	var player_pos = Vector2(player.global_transform.origin.x, player.global_transform.origin.z)
+	if player_pos.distance_to(_prev_player_pos) < 0.1:
+		return
+	_prev_player_pos = player_pos
+	player_pos += interactive_snow_size/2.0;
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd:26]:
```suggestion
	player_pos += interactive_snow_size / 2.0;
```

File: godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+func _ready():
+	get_texture().set_flags(Texture.FLAG_FILTER)
+	var player = get_node(interactive_actor_path)
+	_prev_player_pos = Vector2(player.global_transform.origin.x, player.global_transform.origin.z)
+	var player_pos = _prev_player_pos
+	player_pos += interactive_snow_size/2.0;
+	get_node(interactive_actor_footprint).global_position = (player_pos/ interactive_snow_size) * size
+	var line = get_node("Line2D")
+	line.add_point( (player_pos/ interactive_snow_size) * size)
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd:18]:
```suggestion
	line.add_point((player_pos / interactive_snow_size) * size)
```

File: godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+export var interactive_snow_size: Vector2 = Vector2(32,32)
+
+var _prev_player_pos = Vector2()
+
+func _ready():
+	get_texture().set_flags(Texture.FLAG_FILTER)
+	var player = get_node(interactive_actor_path)
+	_prev_player_pos = Vector2(player.global_transform.origin.x, player.global_transform.origin.z)
+	var player_pos = _prev_player_pos
+	player_pos += interactive_snow_size/2.0;
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd:15]:
```suggestion
	player_pos += interactive_snow_size / 2.0;
```

File: godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd [hunk 8]
```diff
@@ -0,0 +1,34 @@
+extends Viewport
+
+export var interactive_actor_path: NodePath
+export var interactive_actor_footprint: NodePath
+export var line2d: NodePath = "Line2D"
+export var interactive_snow_size: Vector2 = Vector2(32,32)
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd:6]:
```suggestion
export var interactive_snow_size: Vector2 = Vector2(32, 32)
```

File: godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd [hunk 9]
```diff
@@ -0,0 +1,34 @@
+extends Viewport
+
+export var interactive_actor_path: NodePath
+export var interactive_actor_footprint: NodePath
+export var line2d: NodePath = "Line2D"
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd:5]:
```suggestion
export var line2d: NodePath = @"Line2D"
```

**Review by aaronfranke — COMMENTED** [review]
The `.png~` files are likely unintentional and should be deleted.

File: godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+var _prev_player_pos = Vector2()
+
+func _ready():
+	get_texture().set_flags(Texture.FLAG_FILTER)
+	var player = get_node(interactive_actor_path)
+	_prev_player_pos = Vector2(player.global_transform.origin.x, player.global_transform.origin.z)
+	var player_pos = _prev_player_pos
+	player_pos += interactive_snow_size/2.0;
+	get_node(interactive_actor_footprint).global_position = (player_pos/ interactive_snow_size) * size
+	var line = get_node("Line2D")
```
Comment (aaronfranke) [inline godot/Demos/InteractiveSnow/InteractiveSnowViewport.gd:17]:
```suggestion
	var line = get_node(@"Line2D")
```

**Comment by NathanLovato** [comment]
Thanks @aaronfranke but you don't have to bother with individual edits, we use the GDScript formatter.

**Comment by NathanLovato** [comment]
I merged the PR manually in 590a5b4ceb0c97f3f37cf711d8d8faa84eb2e8ef

Could you use jpg for realistic textures? I converted them and halved their resolution, leading to about 30 times smaller size.

In production, we'd want high-res and crisp assets, but for these online demos, I prefer to favor small sizes and quick import times, for people who have slow internet or low-end computers.

And luckily in this case, it doesn't really affect the result visually speaking.

Thanks for your work, thanks @aaronfranke too for checking the PR!

**Pull request closed without merging**